04-onion-routing.md

```diff
@@ -468,7 +468,7 @@ may contain the following TLV fields:
         * [`short_channel_id`:`short_channel_id`]
     1. type: 4 (`next_node_id`)
     2. data:
-        * [`point`:`node_id`]
+        * [`sciddir_or_pubkey`:`node_id`]
```

Comment:
Note for reviewers: this is the important part of that PR. This is a breaking change. But since nobody actually uses onion messages right now, we can afford to make that breaking change now (and avoid introducing a feature bit just for this).

This is necessary for nodes who want to send onion messages to blinded paths without an access to the graph data to resolve an `sciddir` into a `node_id`.

Also, it removes this small chunk from the Bolt 12 PR and thus makes it slightly smaller!

We're curious to know what you all think @rustyrussell @TheBlueMatt @jkczyz @valentinewallace

Reply:
I'm not sure I understand the rationale in the PR description with regards to trampoline. Didn't we include ba2b95a98a415a24c27685a34648c1f64468a4bf in #798 for this purpose? The solution in this PR uses an extra byte per hop compared to that commit, IIUC.

Reply:
The description of the PR isn't very clear, let me try to detail that a bit more. Alice is a mobile wallet user who wants to pay an offer from Bob containing a blinded path that uses an `sciddir` for the introduction node. Alice doesn't have access to the graph, so she cannot resolve the `sciddir` to figure out the introduction node of the blinded path. Alice needs to send an onion message to that blinded path and relies on her LSP for that, so the only thing she can do is create an onion message that uses the following route:

```
Alice ---> LSP -- (sciddir) --> IntroductionNode -- (blinded) --> Bob
```

When creating her onion message path, the only value she can put in the `outgoing_node_id` of the payload for her LSP is the introduction node's `sciddir`, and let the LSP resolve that to a `node_id`. Does that make sense?

> Didn't we include https://github.com/lightning/bolts/commit/ba2b95a98a415a24c27685a34648c1f64468a4bf in https://github.com/lightning/bolts/pull/798 for this purpose?

This is different, because here the LSP and the introduction node aren't necessarily connected (and Alice cannot check that because she doesn't have the graph data). An scid would thus not be enough, because the LSP wouldn't know which node of that channel is the target.

Reply:
> Yea, I mean I still really hate having to look at the network graph to find an onion message first hop, but it saves enough space in the offer that its quite nice.

To be honest, I'm really not a fan of the `sciddir` idea in the first place, I find it really annoying to have to resolve it to a `node_id`...I'm not even convinced that the 24 bytes we're saving are worth the extra complexity, but it looks like LDK and CLN think it's worth it, so we're just following along.

Reply:
> Ugh, can y'all do that as a proprietary phoenix extension for now? At least for the forseeable future phoenix seems like the only wallet that is gonna go without a network graph, even on mobile. I'd rather not introduce any additional sciddir anywhere where we can avoid it, and if things don't evolve into many nodes not having a network graph then this would just be more sciddir resolution that we don't need.

Fair enough, we could do this, we figured this would be useful for other wallets out there and thus having it right now in the spec made sense (if we're introducing `sciddir_or_pubkey` anyway), but we can do our own thing.

I'd like @rustyrussell's opinion on this though before closing this PR!

Reply:
It would have been cleaner to always have this type (it would have been v clean to have inside the onion for the next hop!), but here we are, better late than never.

I think it makes sense for smaller QRs, and I think yes, trampoline will need to understand it.  It's not generally hard to implement, though (but I haven't yet implemented it to be sure, that's today's job!).

Reply:
In the HTLC forwarding case, even if we're using an `sciddir_or_pubkey`, you wouldn't need a graph lookup, it has to be referring to one of _your_ channels?

Reply:
This is not about HTLC forwarding but onion message forwarding which already requires the network graph if you're not already connected to the next node. For HTLC forwarding, we still use an outgoing channel id. We only need the network graph in the case of trampoline which is not what this PR is about.

Reply:
How does this work without a network graph for the case when Alice wants to be paid?  Do the blinded paths in her offer simply use the LSP as the introduction node?

Reply:
For the introduction node, Alice can choose any node to which she knows a path. Even without knowing the full network graph that includes herself and all the nodes she has a channel with.
To use the compact representation, she only needs to know one public channel of the introduction node. In case Alice only has private channels and only knows her own channels, then she can't use this compact representation, she would need her LSP to give her a compact node id (that would require a new message type that would only be useful between a wallet and the LSP).